src/components/BookCard/BookCard.tsx

```diff
@@ -1,13 +1,74 @@
-import type { Book } from '../../domain/books/types'
-
+import { useState } from "react";
+import type { Book } from "../../domain/books/types";
+import classes from "./../../styles/bookcard.module.css";
+import bookpic from "./../../assets/book1.jpg";
+import { AppDispatch } from "../../store/store";
+import { useSelector, useDispatch } from "react-redux";
+import { addToCart, cartSelector } from "../../domain/cart/cartSlice";
+import { CartItem, CartState } from "../../domain/cart/types";
+import {toast } from "react-toastify";
+import "react-toastify/dist/ReactToastify.css";
 /* 
-  We would define the structure of the props in the same file as the component.
-  Since, this component has the same props as Type Book, we would assign it to it props.
+We would define the structure of the props in the same file as the component.
+Since, this component has the same props as Type Book, we would assign it to it props.
 */
 type BookProps = Book;
 
 export const BookCard = (props: BookProps): JSX.Element => {
+  const [togglebutton, setToggleButton] = useState<boolean>(false);
+  const cart = useSelector(cartSelector);
+  const dispatch: AppDispatch = useDispatch();
+
+  function checkCart<T extends CartItem[], U extends Book>(
+    cart: T,
+    props: U
+  ): boolean {
+    for (let c of cart) {
+      if (c.id === props.id) return false;
+    }
+    return true;
+  }
+  const onAddClickHandler = () => {
+    if (!togglebutton) {
+      if (checkCart(cart, props)) {
```

Comment:
This condition is used multiple times in this file with the same parameters. Instead call it once and cache the result in a variable.
`const cartHasBook = checkCart(id)`